TESTS/mbedmicro-rtos-mbed/semaphore/main.cpp

```diff
@@ -167,6 +164,97 @@ void test_timeout()
     t.join();
     TEST_ASSERT_UINT32_WITHIN(5000, 30000, timer.read_us());
 }
+#endif
+
+void test_ticker_release(struct test_data *data)
+{
+    osStatus res;
+    data->data++;
+    res = data->sem->release();
+    TEST_ASSERT_EQUAL(osOK, res);
+}
+
+/** Test semaphore acquire
+
+    Given a semaphore with no tokens available and ticker with the callback registered
+    when callbacks update the test data and release semaphore
+    which will unblock the main thread which is blocked on semaphore acquire.
+ */
+void test_semaphore_acquire()
+{
+    Semaphore sem(0);
+    struct test_data data;
+
+    data.sem = &sem;
+    data.data = 0;
+    Ticker t1;
+    t1.attach_us(callback(test_ticker_release, &data), 3000);
+    sem.acquire();
+    t1.detach();
+
+    TEST_ASSERT_EQUAL(1, data.data);
+}
+
+void test_ticker_try_acquire(Semaphore *sem)
+{
+    osStatus res;
+    res = sem->try_acquire();
+    TEST_ASSERT_FALSE(res);
+}
+
+/** Test semaphore try acquire
+
+    Given a semaphore with no tokens available and ticker with the callback registered
+    when callbacks try to acquire the semaphore will fail.
+ */
+void test_semaphore_try_acquire()
+{
+    Semaphore sem(0);
+    Ticker t1;
+    t1.attach_us(callback(test_ticker_try_acquire, &sem), 3000);
+    ThisThread::sleep_for(3);
+    t1.detach();
+}
+
+
+/** Test semaphore try timeout
+
+    Given a semaphore with no tokens available
+    when the main thread calls @a wait on the semaphore with try timeout of 5ms
+    then the main thread waits for 5ms and timeouts after
+ */
+void test_semaphore_try_timeout()
+{
+    Semaphore sem(0);
+    bool res;
+    res = sem.try_acquire_for(5);
+    TEST_ASSERT_FALSE(res);
+}
+
+
+void test_ticker_semaphore_release(struct Semaphore *sem)
+{
+    osStatus res;
+    res = sem->release();
+    TEST_ASSERT_EQUAL(osOK, res);
+}
+
+/** Test semaphore try acquire timeout
+
+    Given a semaphore with no tokens available and ticker with the callback registered
+    when callbacks release the semaphore will unblock the main thread
+    which is waiting for semaphore with a timeout.
+ */
+void test_semaphore_try_acquire_timeout()
+{
+    Semaphore sem(0);
+    bool res;
+    Ticker t1;
+    t1.attach_us(callback(test_ticker_semaphore_release, &sem), 3000);
+    res = sem.try_acquire_for(30);
```

Comment:
`sem.try_acquire_for(10)` should be enough waiting and speeds up the test if case of failure.